File: structsCPP/LinkedList/LinkedList.cpp
#include "LinkedList.h"

#include <iostream>

LinkedList::LinkedList() {
    this->head = nullptr;
    this->tail = nullptr;
    this->length = 0;
}

LinkedList::LinkedList(const int& value) {
    Node* newNode = new Node(value);
    this->head = newNode;
    this->tail = newNode;
    this->length = 1;
}

LinkedList::~LinkedList() {
    Node* temp = this->head;
    while (this->head) {
        this->head = this->head->next;
        delete temp;
        temp = this->head;
    }
    this->tail = nullptr;
    this->length = 0;
}

void LinkedList::append(const int& value) {
    Node* newNode = new Node(value);

    if (this->length == 0) {
        this->head = newNode;
        this->tail = newNode;
    }
    else {
        this->tail->next = newNode;
        this->tail = newNode;
    }
    this->length++;
}

void LinkedList::prepend(const int& value) {
    Node* newNode = new Node(value);

    if (this->length == 0) {
        this->head = newNode;
        this->tail = newNode;
    }
    else {
        newNode->next = this->head;
        this->head = newNode;
    }
    this->length++;
}

bool LinkedList::insert(const int& index, const int& value) {
    if (index < 0 || index > this->length) { return false; }

    if (index == 0) {
        this->prepend(value);
        return true;
    }

    if (index == this->length) {
        this->append(value);
        return true;
    }

    Node* newNode = new Node(value);
    Node* temp = this->getNode(index - 1);

    newNode->next = temp->next;
    temp->next = newNode;

    this->length++;
    return true;
}

bool LinkedList::setNode(const int& index, const int& value) {
    Node* temp = this->getNode(index);
    if (temp) {
        temp->value = value;
        return true;
    }
    return false;
}

void LinkedList::deleteNode(const int& index) {
    if (index < 0 || index >= this->length) return;

    if (index == 0) return this->removeFirst();

    if (index == this->length - 1) return this->removeLast();

    Node* prevNode = this->getNode(index - 1);
    Node* temp = prevNode->next;
    prevNode->next = temp->next;

    delete temp;
    this->length--;
}


void LinkedList::removeFirst() {
    if (this->length == 0) {
        return;
    }

    Node* temp = this->head;

    if (this->length == 1) {
        this->head = nullptr;
        this->tail = nullptr;
    }
    else {
        this->head = this->head->next;
    }

    delete temp;
    this->length--;
}


void LinkedList::removeLast() {
    if (this->length == 0) { return; }

    Node* temp = this->head;

    if (this->length == 1) {
        this->head = nullptr;
        this->tail = nullptr;
    }
    else {
        Node* prevNode = temp;
        while (temp->next != nullptr) {
            prevNode = temp;
            temp = temp->next;
        }
        tail = prevNode;
        tail->next = nullptr;
    }

    delete temp;
    this->length--;
}

void LinkedList::printList() const {
    Node* temp = this->head;
    if (!temp) {
        std::cout << "Empty List!\n";
        return;
    }

    while (temp != nullptr) {
        std::cout << temp->value;
        temp = temp->next;
        if (temp != nullptr) {
            std::cout << " -> ";
        }
    }
}

Node* LinkedList::getNode(const int& index) const {
    if (index < 0 || index >= this->length) return nullptr;

    Node* temp = this->head;

    for (int i = 0; i < index; i++) {
        temp = temp->next;
    }

    return temp;
}

Node* LinkedList::getHead() const {
    return this->head;
}

Node* LinkedList::getTail() const {
    return this->tail;
}

int LinkedList::getLength() const {
    return this->length;
}

void LinkedList::reverse() {
    Node* temp = this->head;
    this->head = this->tail;
    this->tail = temp;

    Node* prevNode = nullptr;
    Node* nextNode = temp->next;

    for (int i = 0; i < length; i++) {
        nextNode = temp->next;
        temp->next = prevNode;
        prevNode = temp;
        temp = nextNode;
    }

}

Node* LinkedList::findMiddleNode() const {
    if(this->length == 0) return nullptr;

    const int half = this->length / 2;
    Node* temp = this->head;

    for (int i = 0; i < half; ++i) {
        temp = temp->next;
    }
    return temp;
}

Node* LinkedList::middle() const {
    Node* slow = this->head;
    Node* fast = this->head;

    if (slow == nullptr) return nullptr;

    if (slow->next == nullptr) return slow;

    while(fast != nullptr) {
        if (fast->next == nullptr) break;
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow;
}

bool LinkedList::hasLoop() const {
    if (this->head == nullptr) return false;
    if (this->head->next == nullptr) return false;

    Node* slow = this->head;
    Node* fast = this->head;

    while (fast != nullptr) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) return true;
    }

    return false;
}

File: structsPYTHON/Lists/LinkedList/LinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self, value):
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def append(self, value) -> None:
        new_node = Node(value)

        if self.length == 0:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

        self.length += 1

    def prepend(self, value) -> None:
        new_node = Node(value)

        if self.length == 0:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node

        self.length += 1

    def insert(self, index, value) -> bool:
        if index == 0:
            self.prepend(value)
            return True
        if index == self.length - 1:
            self.append(value)
            return True

        prev_node = self.get_node(index - 1)
        if prev_node is not None:
            new_node = Node(value)
            new_node.next = prev_node.next
            prev_node.next = new_node
            self.length += 1
            return True

        return False

    def pop_first(self) -> Node | None:
        if self.length == 0:
            return None

        temp_node = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
            temp_node.next = None

        self.length -= 1
        return temp_node

    def pop(self) -> Node | None:
        if self.length == 0:
            return None

        temp_node = self.head

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            while temp_node.next.next is not None:
                temp_node = temp_node.next

            self.tail = temp_node
            temp_node = temp_node.next
            self.tail.next = None

        self.length -= 1
        return temp_node

    def remove_node(self, index) -> bool: # deleteNode
        if index == 0:
            temp_node = self.pop_first()
            del temp_node
            return True
        if index == self.length - 1:
            temp_node = self.pop()
            del temp_node
            return True

        previous_node = self.get_node(index - 1)
        if previous_node is not None:
            temp_node = previous_node.next
            previous_node.next = temp_node.next
            temp_node.next = None
            del temp_node
            self.length -= 1
            return True

        return False

    def print(self) -> None:
        if self.length == 0:
            print("List is empty")
            return

        temp = self.head

        while temp:
            print(f"{temp.value}")
            temp = temp.next

    def set_value(self, index, value) -> bool:
        pre_updated_node = self.get_node(index)
        if pre_updated_node is not None:
            pre_updated_node.value = value
            return True
        return False

    def get_node(self, index) -> Node | None:
        if self.length == 0:
            return None
        if index < 0 or index >= self.length:
            return None

        count = 0
        temp_node = self.head
        while count < index:
            temp_node = temp_node.next
            count += 1

        return temp_node

    def get_head(self) -> int:
        return self.head.value

    def get_tail(self) -> int:
        return self.tail.value

    def get_length(self) -> int:
        return self.length


